### PR TITLE
ISSUE-12: Reduce peak memory usage when building

### PR DESCRIPTION
**Pull request by klaussfreire**
Should improve the situation in issue #12 : 
* Concatenate without building a temporary numpy array duplicating the contents
* Dereference temporary arrays earlier
* Coalesce big parts list into smaller lists of big parts to reduce memory fragmentation and improve deduplication
* Specialize the algorithm used by numpy.unique for our purposes, to avoid the RAM wastage induced by numpy's generality. In particular:
  - Don't use argsort, the indices array weighs a lot, use in-place sort (since we feed it scratch buffers, we can alter the input)
  - Compute flags using sliced views of the input array
  - Don't copy the array if all flags are true, just return it
  - Don't compute indexes out of the flags, just use the flags

### CODE REVIEW

File: sharedbuffers/mapped_struct.py [hunk 1]
```diff
@@ -2622,6 +2623,21 @@ def _merge_all(parts, dtype):
         del parts
         return _merge_all(nparts, dtype)
 
+@cython.cfunc
+@cython.locals(parts = list, discard_duplicate_keys = cython.bint, discard_duplicates = cython.bint)
+def _discard_duplicates(apart, struct_dt, discard_duplicate_keys, discard_duplicates):
+    if discard_duplicate_keys or discard_duplicates:
+        # What numpy.unique does, but using MUCH less RAM
+        vpart = apart.view(struct_dt)
+        vpart.sort(0)
```
Comment (zeitos) [inline sharedbuffers/mapped_struct.py:2632]: what's 0?

File: sharedbuffers/mapped_struct.py [hunk 2]
```diff
@@ -2622,6 +2623,21 @@ def _merge_all(parts, dtype):
         del parts
         return _merge_all(nparts, dtype)
 
+@cython.cfunc
+@cython.locals(parts = list, discard_duplicate_keys = cython.bint, discard_duplicates = cython.bint)
+def _discard_duplicates(apart, struct_dt, discard_duplicate_keys, discard_duplicates):
+    if discard_duplicate_keys or discard_duplicates:
+        # What numpy.unique does, but using MUCH less RAM
+        vpart = apart.view(struct_dt)
```
Comment (zeitos) [inline sharedbuffers/mapped_struct.py:2631]: where can I find the documentation for view?

Comment (klaussfreire) [reply sharedbuffers/mapped_struct.py:2631]: https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.view.html

File: sharedbuffers/mapped_struct.py (hunk 1 above)
Comment (klaussfreire) [reply sharedbuffers/mapped_struct.py:2632]: The first axis (first dimension)

Comment (zeitos) [reply sharedbuffers/mapped_struct.py:2632]: talked in person

**Review by zeitos — APPROVED** [review]

**Review by tzulberti-jampp — APPROVED** [review]

**Pull request merged**